samplers/jaegerremote/per_operation_sampler.go

```diff
@@ -0,0 +1,60 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jaegerremote
+
+import (
+	"fmt"
+	"strings"
+
+	jaeger_api_v2 "go.opentelemetry.io/contrib/samplers/jaegerremote/internal/proto-gen/jaeger-idl/proto/api_v2"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+type perOperationSampler struct {
+	defaultSampler trace.Sampler
+	operationMap   map[string]trace.Sampler
+}
+
+var _ trace.Sampler = &perOperationSampler{}
+
+func (p *perOperationSampler) ShouldSample(parameters trace.SamplingParameters) trace.SamplingResult {
+	sampler := p.defaultSampler
+	if s, ok := p.operationMap[parameters.Name]; ok {
+		sampler = s
+	}
+	return sampler.ShouldSample(parameters)
+}
+
+func (p *perOperationSampler) Description() string {
+	var ops []string
+	for op, sampler := range p.operationMap {
+		ops = append(ops, fmt.Sprintf("%s:%s", op, sampler.Description()))
+	}
+	mapStr := strings.Join(ops, ",")
+
+	return fmt.Sprintf("PerOperationSampler{default=%s,perOperation={%v}}", p.defaultSampler.Description(), mapStr)
+}
+
+func newPerOperationSampler(strategies *jaeger_api_v2.PerOperationSamplingStrategies) trace.Sampler {
```

Comment:
Per-operation sampler in Jaeger uses two combined samplers per service/operation pair: a probabilistic sampler and a rate-limited sampler to ensure guaranteed minimum rate of traces. See the code here: https://github.com/jaegertracing/jaeger-client-go/blob/669feb8e260bd722003cd14ea6b0a808b9d8e533/sampler.go#L293

I am curious - why re-implement everything from scratch instead of simply cloning the above code from Jaeger and just adapting it to OTEL Sampler API slightly?

Reply:
I'd also be in favor of wrapping the existing implementations.  To the extent that we're able to re-use them since we're using the same language we probably should.  

@yurishkuro given the recent announcement of the deprecation of the Jaeger client libraries can we rely on this remaining available or should we copy the implementation into our own packages?

Reply:
While the Jaeger libraries' code will remain on GitHub and distros, it was not designed to allow internal components like samplers to be imported directly into other projects. I would recommend taking a copy, especially since there's not going to be ongoing improvements to the original code.

Reply:
IIRC the APIs didn't align really well so I couldn't just write a wrapper around the Jaeger code. Additionally, I didn't want to just link into the Jaeger codebase as that would blow up the go dependencies.

Agree copying will be better than trying to link. Just copying all code over and committing it as if it's mine feels weird, is there any comment or notice that is typically added when adopting code from another project?

Reply:
> is there any comment or notice that is typically added when adopting code from another project?

the accepted practice is to copy copyright bylines from the original code, and add OTEL on top. E.g. in Jaeger you will often find these:
```
// Copyright (c) 2019 The Jaeger Authors.
// Copyright (c) 2017 Uber Technologies, Inc.
```

Reply:
> IIRC the APIs didn't align really well

the APIs are actually a small part of the code (sampling API was just one function originally), implementation of remote polling, parsing, etc. is much larger and could remain almost intact.